gcloud/src/main/java/org/jimmutable/gcloud/search/IndexManager.java

```diff
@@ -0,0 +1,167 @@
+				index.delete(index_id.getSimpleValue());
+				logger.info(new LogSupplier("Deleted documentId %s in indexId %s", index_id.getSimpleValue(),
+						document_id.getSimpleValue()));
+				break;
+			} catch (RuntimeException e) {
+				logger.log(Level.SEVERE, e, new LogSupplier("Failed to delete documentId %s in indexId %s"));
+				Thread.sleep(delay * 1000);
+				delay *= 2; // easy exponential backoff
+			}
+		}
+	}
+
+	/**
+	 * Performs a deletion on a DocumentId within an IndexId and will retry up to
+	 * three times in the even a runtimeException is thrown.
+	 * 
+	 * @param indexable
+	 *            Indexable
+	 * @throws InterruptedException
+	 */
+	public static void delete(Indexable indexable) throws InterruptedException {
+		delete(indexable.getSimpleSearchIndexId(), indexable.getSimpleSearchDocumentId());
+	}
+
+	/**
+	 * 
+	 * Creates a Document using DocumentWriter.createDocument() and puts it into the
+	 * index. This will retry up to three times if the put fails from a system
+	 * related issue.
+	 * 
+	 * @param indexable
+	 *            Indexable
+	 * @throws InterruptedException
+	 */
+	public static void upsert(Indexable indexable) throws InterruptedException {
+
+		Index index = getIndex(indexable.getSimpleSearchIndexId());
+
+		DocumentWriter writer = new DocumentWriter(indexable.getSimpleSearchDocumentId());
+		indexable.writeSearchDocument(writer);
+
+		final int maxRetry = 3;
+		int attempts = 0;
+		int delay = 2;
+		while (true) {
+			try {
+				index.put(writer.createDocument());
+				logger.info(new LogSupplier("Upserted documentId %s into indexId %s",
+						indexable.getSimpleSearchDocumentId().getSimpleValue(),
+						indexable.getSimpleSearchIndexId().getSimpleValue()));
+			} catch (PutException e) {
+				if ((StatusCode.TRANSIENT_ERROR.equals(e.getOperationResult().getCode())
+						|| StatusCode.CONCURRENT_TRANSACTION_ERROR.equals(e.getOperationResult().getCode())
+						|| StatusCode.INTERNAL_ERROR.equals(e.getOperationResult().getCode())
+						|| StatusCode.TIMEOUT_ERROR.equals(e.getOperationResult().getCode()))
+						&& ++attempts < maxRetry) { // retrying
+
+					logger.log(Level.WARNING, e, new LogSupplier(
+							"Upsert for documentId %s into indexId %s failed with status code %s. Retrying in %d seconds...",
+							indexable.getSimpleSearchDocumentId().getSimpleValue(),
+							indexable.getSimpleSearchIndexId().getSimpleValue(),
+							e.getOperationResult().getCode().name(), delay));
+
+					Thread.sleep(delay * 1000);
+					delay *= 2; // easy exponential backoff
+					continue;
+				} else {
+
+					logger.log(Level.SEVERE, e,
+							new LogSupplier("Upsert for documentId %s into indexId %s failed with status code %s",
+									indexable.getSimpleSearchDocumentId().getSimpleValue(),
+									indexable.getSimpleSearchIndexId().getSimpleValue(),
+									e.getOperationResult().getCode().name()));
+
+					// PutException is a RuntimeException
+					throw e;
+				}
+			}
+			break;
+		}
+	}
+
+}
```

Comment:
Validate input params

Don't throw an exception, return a boolean and catch all exceptions